services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -410,6 +415,11 @@ public RefreshResult refreshAccessToken(KeycloakSession session, UriInfo uriInfo
 
         AccessTokenResponse res = responseBuilder.build();
 
+        // KEYCLOAK-15169 OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)
+        if (clientConfig.isDPoPEnabled()) {
```

Comment:
[The design documents](https://github.com/keycloak/keycloak-community/blob/abdb40b888e9c7689239679ca30cc7a6a4376c2e/design/dpop.md#relation-to-other-token-binding-mechanisms) said that ether MTLS or DPoP should be applied per client, not both. The current codes accept a client that MTLS and DPoP are both applied to. If following the design document, some methanism is needed to only allow ether MTLS or DPoP to be applied per client.

Reply:
I agree that MTLS and DPoP (and other hypothetical binding mechanisms) should be mutually exclusive; I think we're also aligned with the OAuth WG on that. However, to properly implement mutual exclusivity, I'm afraid we first need to unify MTLS and DPoP via implementing some common interfaces (be it Token Binding SPI or Client Policies or something else). I don't think we can achieve the same without refactoring both MTLS and DPoP code. That said, such change would not be directly relevant to DPoP support.

I'd suggest that we implement mutual exclusivity during the next DPoP development phase, WDYT? This will be reflected  in the design doc of course.

Reply:
@dteleguin Yes, I agree with your suggestion.